### PR TITLE
DB comparisons, and switching to Badger

### PR DESCRIPTION
**Pull request by hasyimibhar**
This PR explores performance differences between different KV stores. You can run the benchmarks yourself, which you can find in the `benchmarks` folder. Each benchmark produces a CSV output so you can easily chart it using R/Python/Excel to visualize the result.

Major changes:

- `WriteBatch.Put` now returns an error. `WriteBatch` also does not guarantee all of the batched operations are not committed if it's not explicitly committed with `KV.CommitWriteBatch`, as the underlying database may choose to commit the write batch earlier. What `WriteBatch` guarantees is that if N operations are batched before being committed to disk, the underlying database will write it to disk in at least 1 batch, not exactly 1 batch.

TODO:
- [x] Set Badger as KV for Wavelet

### CODE REVIEW

**Comment by hasyimibhar** [comment]
```sh
$ go test -bench=BenchmarkAccountsCommit -run=BenchmarkAccountsCommit -benchtime=5x -timeout=300m
goos: linux
goarch: amd64
pkg: github.com/perlin-network/wavelet
BenchmarkAccountsCommit/badger/100MB-8         	       5	1889863337 ns/op
BenchmarkAccountsCommit/bbolt/100MB-8          	       5	1827996310 ns/op
BenchmarkAccountsCommit/level/100MB-8          	       5	7194924344 ns/op
BenchmarkAccountsCommit/badger/200MB-8         	       5	3384995602 ns/op
BenchmarkAccountsCommit/bbolt/200MB-8          	       5	3533082702 ns/op
BenchmarkAccountsCommit/level/200MB-8          	       5	13936901039 ns/op
BenchmarkAccountsCommit/badger/300MB-8         	       5	4836044848 ns/op
BenchmarkAccountsCommit/bbolt/300MB-8          	       5	5183316483 ns/op
BenchmarkAccountsCommit/level/300MB-8          	       5	20943055189 ns/op
BenchmarkAccountsCommit/badger/400MB-8         	       5	6421188462 ns/op
BenchmarkAccountsCommit/bbolt/400MB-8          	       5	7359878551 ns/op
BenchmarkAccountsCommit/level/400MB-8          	       5	28137950461 ns/op
BenchmarkAccountsCommit/badger/500MB-8         	       5	8464443560 ns/op
BenchmarkAccountsCommit/bbolt/500MB-8          	       5	8981183222 ns/op
BenchmarkAccountsCommit/level/500MB-8          	       5	34608603426 ns/op
BenchmarkAccountsCommit/badger/600MB-8         	       5	7793629533 ns/op
BenchmarkAccountsCommit/bbolt/600MB-8          	       5	8813665213 ns/op
BenchmarkAccountsCommit/badger/700MB-8         	       5	9165851906 ns/op
BenchmarkAccountsCommit/bbolt/700MB-8          	       5	10251347187 ns/op
BenchmarkAccountsCommit/badger/800MB-8         	       5	9776101783 ns/op
BenchmarkAccountsCommit/bbolt/800MB-8          	       5	11754434801 ns/op
BenchmarkAccountsCommit/badger/900MB-8         	       5	11346871333 ns/op
BenchmarkAccountsCommit/bbolt/900MB-8          	       5	15850786563 ns/op
BenchmarkAccountsCommit/badger/1000MB-8        	       5	18770368991 ns/op
BenchmarkAccountsCommit/bbolt/1000MB-8         	       5	15488524497 ns/op
PASS
ok  	github.com/perlin-network/wavelet	1649.641s
```

**Comment by hasyimibhar** [comment]
Visualized output of `go run ./benchmarks tree` (it benchmarks a single `tree.Commit` call):

## Below 100MB

benchmark_sub100

## Above 100MB

benchmark_above100

## How to reproduce using R and ggplot2

```r
require('ggplot2')
data <- read.csv('/path/to/output.csv')

# Below 100MB
ggplot(data[data$size < 102400,], aes(x=size/1024, y=time/1000000000, color=db, shape=db)) +
    geom_point(size=2) + facet_grid(rows=vars(hd_type), scales='free_y') +
    xlab('Commit size (MB)') + ylab('Time (s)') +
    labs(title='tree.Commit benchmark on various KV databases', subtitle='HDD vs SSD') + theme_light() +
    theme(plot.title = element_text(hjust = 0.5), plot.subtitle = element_text(hjust = 0.5)) +
    scale_x_log10()


# Above 100MB
ggplot(data[data$size > 102400,], aes(x=size/1024, y=time/1000000000, color=db, shape=db)) +
    geom_point(size=2) + facet_grid(rows=vars(hd_type), scales='free_y') +
    xlab('Commit size (MB)') + ylab('Time (s)') +
    labs(title='tree.Commit benchmark on various KV databases', subtitle='HDD vs SSD') + theme_light() +
    theme(plot.title = element_text(hjust = 0.5), plot.subtitle = element_text(hjust = 0.5)) +
    scale_x_log10()
```

File: store/badger.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	batch *badger.WriteBatch
+}
+
+func (b *badgerWriteBatch) Put(key, value []byte) {
+	k := append([]byte{}, key...)
+	v := append([]byte{}, value...)
+	err := b.batch.Set(k, v)
+	if err != nil {
+		panic(err)
+	}
```
Comment (AhmadMuzakkir) [inline store/badger.go]: Not sure if this is the best way to handle the error returned by Set()

It seems that if the batch is too big, the Set() will call commit() internally. So, the error might be coming from commit().

Comment (hasyimibhar) [reply store/badger.go]: Good catch. Looks like `WriteBatch.Put` needs to return an error. And it must be made clear that `WriteBatch` does not guarantee all of the batched operations are not committed if it's not explicitly committed, as the underlying database may choose to commit the write batch earlier.

**Comment by rkeene** [comment]
```
# github.com/perlin-network/wavelet/store
store/testutil.go:61:3: cannot use badger (type *badgerKV) as type KV in return argument:
	*badgerKV does not implement KV (missing Dir method)
store/testutil.go:75:3: cannot use bbolt (type *bboltKV) as type KV in return argument:
	*bboltKV does not implement KV (missing Dir method)
# github.com/perlin-network/wavelet/store [github.com/perlin-network/wavelet/store.test]
store/testutil.go:61: cannot use badger (type *badgerKV) as type KV in return argument:
	*badgerKV does not implement KV (missing Dir method)
store/testutil.go:75: cannot use bbolt (type *bboltKV) as type KV in return argument:
	*bboltKV does not implement KV (missing Dir method)
```

**Comment by hasyimibhar** [comment]
Closing this as it's stale. It has been reincarnated as #326.

**Pull request closed without merging**